hal/armv8m/nrf/hal.c

```diff
@@ -0,0 +1,200 @@
+extern void timer_done(void);
+
+/* Console */
+extern void console_init(void);
+
+/* Interrupts */
+extern void interrupts_init(void);
+
+
+void hal_init(void)
+{
+	hal_cpuInit();
+
+	_nrf91_init();
+	interrupts_init();
+
+	timer_init();
+	console_init();
+
+	hal_common.entry = (addr_t)-1;
+}
+
+
+void hal_done(void)
+{
+	timer_done();
+}
+
+
+void hal_syspageSet(hal_syspage_t *hs)
+{
+	hal_common.hs = hs;
+}
+
+
+const char *hal_cpuInfo(void)
+{
+	return "Cortex-M33 nRF9160";
+}
+
+
+addr_t hal_kernelGetAddress(addr_t addr)
+{
+	return addr;
+}
+
+
+void hal_kernelGetEntryPointOffset(addr_t *off, int *indirect)
+{
+	*off = sizeof(unsigned);
+	*indirect = 1;
+}
+
+
+void hal_kernelEntryPoint(addr_t addr)
+{
+	hal_common.entry = addr;
+}
+
+
+int hal_memoryAddMap(addr_t start, addr_t end, u32 attr, u32 mapId)
+{
+	return 0;
+}
+
+
+static void hal_getMinOverlappedRange(addr_t start, addr_t end, mapent_t *entry, mapent_t *minEntry)
+{
+	if ((start < entry->end) && (end > entry->start)) {
+		if (start > entry->start) {
+			entry->start = start;
+		}
+
+		if (end < entry->end) {
+			entry->end = end;
+		}
+
+		if (entry->start < minEntry->start) {
+			minEntry->start = entry->start;
+			minEntry->end = entry->end;
+			minEntry->type = entry->type;
+		}
+	}
+}
+
+
+int hal_memoryGetNextEntry(addr_t start, addr_t end, mapent_t *entry)
+{
+	int i;
+	mapent_t tempEntry, minEntry;
+
+	static const mapent_t entries[] = {
+		{ .start = (addr_t)__init_start, .end = (addr_t)__init_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__text_start, .end = (addr_t)__etext, .type = hal_entryTemp },
+		{ .start = (addr_t)__rodata_start, .end = (addr_t)__rodata_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__init_array_start, .end = (addr_t)__init_array_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__fini_array_start, .end = (addr_t)__fini_array_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__ramtext_start, .end = (addr_t)__ramtext_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__data_start, .end = (addr_t)__data_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__bss_start, .end = (addr_t)__bss_end, .type = hal_entryTemp },
+		{ .start = (addr_t)__heap_base, .end = (addr_t)__heap_limit, .type = hal_entryTemp },
+		{ .start = (addr_t)__stack_limit, .end = (addr_t)__stack_top, .type = hal_entryTemp },
+	};
+
+	if (start == end) {
+		return -1;
+	}
+
+	hal_memset(&tempEntry, 0, sizeof(tempEntry));
+	hal_memset(&minEntry, 0, sizeof(minEntry));
+	minEntry.start = (addr_t)-1;
+
+	/* Syspage entry */
+	tempEntry.start = (addr_t)hal_common.hs;
+	tempEntry.end = (addr_t)__heap_limit;
+	tempEntry.type = hal_entryReserved;
+	hal_getMinOverlappedRange(start, end, &tempEntry, &minEntry);
+
+	for (i = 0; i < sizeof(entries) / sizeof(entries[0]); ++i) {
+		if (entries[i].start >= entries[i].end) {
+			continue;
+		}
+		hal_memcpy(&tempEntry, &entries[i], sizeof(mapent_t));
+		hal_getMinOverlappedRange(start, end, &tempEntry, &minEntry);
+	}
```

Comment:
Look at this commit
https://github.com/phoenix-rtos/plo/commit/52efcba4df06aea5ffc34a668e171efc3427465c

Reply:
Thanks! Applied.